applications/OptimizationApplication/python_scripts/algorithms/algorithm_gradient_projection.py

```diff
@@ -1,120 +1,227 @@
+
+        self.master_control = MasterControl() # Need to fill it with controls
+
+        for control_name in parameters["controls"].GetStringArray():
+            control = optimization_problem.GetControl(control_name)
+            self.master_control.AddControl(control)
+
+
+        settings = parameters["settings"]
+        settings.ValidateAndAssignDefaults(self.GetDefaultParameters()["settings"])
+
+        self.echo_level = settings["echo_level"].GetInt()
+
+        ComponentDataView("algorithm", self._optimization_problem).SetDataBuffer(self.GetMinimumBufferSize())
+
+        self.__convergence_criteria = CreateConvergenceCriteria(settings["conv_settings"], self._optimization_problem)
+        self.__line_search_method = CreateLineSearch(settings["line_search"], self._optimization_problem)
+
+        self.__objective = StandardizedObjective(parameters["objective"], self.master_control, self._optimization_problem)
+        self.__constraints_list: 'list[StandardizedConstraint]' = []
+        for constraint_param in parameters["constraints"]:
+            constraint = StandardizedConstraint(constraint_param, self.master_control, self._optimization_problem)
+            self.__constraints_list.append(constraint)
+        self.__control_field = None
+        self.__obj_val = None
+
+        default_linear_solver_settings = Kratos.Parameters("""{
+            "solver_type": "LinearSolversApplication.dense_col_piv_householder_qr"
+        }""")
+        settings["linear_solver_settings"].ValidateAndAssignDefaults(default_linear_solver_settings)
+        self.linear_solver = ConstructSolver(settings["linear_solver_settings"])
+        self.correction_size = settings["correction_size"].GetDouble()
+
+    def GetMinimumBufferSize(self) -> int:
+        return 2
+
+    def Check(self):
+        pass
+    
+    @time_decorator()
+    def Initialize(self):
+        self.converged = False
+        self.__obj_val = None
+        self.__objective.Initialize()
+        self.__objective.Check()
+        CallOnAll(self.__constraints_list, StandardizedConstraint.Initialize)
+        CallOnAll(self.__constraints_list, StandardizedConstraint.Check)
+        self.master_control.Initialize()
+        self.__control_field = self.master_control.GetControlField()
+        self.algorithm_data = ComponentDataView("algorithm", self._optimization_problem)
+
+    @time_decorator()
+    def Finalize(self):
+        pass
+    
+    @time_decorator()
+    def ComputeSearchDirection(self, obj_grad: KratosOA.CollectiveExpression, constr_grad: 'list[KratosOA.CollectiveExpression]') -> KratosOA.CollectiveExpression:
+        active_constraints_list = [self.__constraints_list[i] for i in range(len(self.__constraints_list)) if self.__constr_value[i] >= 0.0]
+        number_of_active_constraints = len(active_constraints_list)
+        if not number_of_active_constraints:
+            search_direction = obj_grad.Clone() * -1.0
+            correction = obj_grad.Clone() * 0.0
```

Comment:
```suggestion
            search_direction = obj_grad * -1.0
            correction = obj_grad * 0.0
```

When you do an arithmetic operation, it anyways needs to create a copy.

Reply:
So, search_direction is detached from obj_grad?